Cython/Compiler/Code.py

```diff
@@ -2060,131 +2062,94 @@ def entry_as_pyobject(self, entry):
-        X = 'X' if null_check else ''
+    def put_var_decref(self, entry, **kwds):
+        self.putln(entry.type.generate_decref(entry.cname, **kwds))
 
-        if clear:
-            if clear_before_decref:
-                if not nanny:
-                    X = ''  # CPython doesn't have a Py_XCLEAR()
-                self.putln("%s_%sCLEAR(%s);" % (prefix, X, cname))
-            else:
-                self.putln("%s_%sDECREF(%s); %s = 0;" % (
-                    prefix, X, self.as_pyobject(cname, type), cname))
-        else:
-            self.putln("%s_%sDECREF(%s);" % (
-                prefix, X, self.as_pyobject(cname, type)))
+    def put_var_xdecref(self, entry, **kwds):
+        self.putln(entry.type.generate_xdecref(entry.cname, **kwds))
 
-    def put_decref_set(self, cname, rhs_cname):
-        self.putln("__Pyx_DECREF_SET(%s, %s);" % (cname, rhs_cname))
+    def put_var_decref_clear(self, entry, **kwds):
+        self._put_var_decref_clear(entry, null_check=False, **kwds)
 
-    def put_xdecref_set(self, cname, rhs_cname):
-        self.putln("__Pyx_XDECREF_SET(%s, %s);" % (cname, rhs_cname))
+    def put_var_xdecref_clear(self, entry, **kwds):
+        self._put_var_decref_clear(entry, null_check=True, **kwds)
 
-    def put_var_decref(self, entry):
-        if entry.type.is_pyobject:
-            self.putln("__Pyx_XDECREF(%s);" % self.entry_as_pyobject(entry))
+    def _put_var_decref_clear(self, entry, null_check, **kwds):
+        f = getattr(entry.type, "generate_%sdecref_clear" %
+                                (null_check and "x" or ""))
+        self.putln(f(entry.cname,
+                     clear_before_decref=entry.in_closure,
+                     **kwds))
 
-    def put_var_xdecref(self, entry, nanny=True):
-        if entry.type.is_pyobject:
-            if nanny:
-                self.putln("__Pyx_XDECREF(%s);" % self.entry_as_pyobject(entry))
-            else:
-                self.putln("Py_XDECREF(%s);" % self.entry_as_pyobject(entry))
-
-    def put_var_decref_clear(self, entry):
-        self._put_var_decref_clear(entry, null_check=False)
-
-    def put_var_xdecref_clear(self, entry):
-        self._put_var_decref_clear(entry, null_check=True)
-
-    def _put_var_decref_clear(self, entry, null_check):
-        if entry.type.is_pyobject:
-            if entry.in_closure:
-                # reset before DECREF to make sure closure state is
-                # consistent during call to DECREF()
-                self.putln("__Pyx_%sCLEAR(%s);" % (
-                    null_check and 'X' or '',
-                    entry.cname))
-            else:
-                self.putln("__Pyx_%sDECREF(%s); %s = 0;" % (
-                    null_check and 'X' or '',
-                    self.entry_as_pyobject(entry),
-                    entry.cname))
+    def get_var_nullcheck(self, entry):
+        # we usually want to put this into an is statement so it's a get,
+        # not a put
```

Comment:
I don't think we use `get_…` elsewhere. It's just `error_goto_if()`, for example.